layer2-state-provider/history_commitment_provider.go

```diff
@@ -0,0 +1,281 @@
+	if len(validatedHeights) == 1 {
+		hashes, hashesErr := p.l2MessageStateCollector.L2MessageStatesUpTo(ctx, Height(fromMessageNumber), upToHeight, batch)
+		if hashesErr != nil {
+			return emptyCommit, hashesErr
+		}
+		return commitments.New(hashes)
+	}
+
+	// Next, computes the exact start point of where we need to execute
+	// the machine from the inputs, and figures out in what increments we need to do so.
+	machineStartIndex := p.computeMachineStartIndex(validatedHeights)
+
+	// We compute the stepwise increments we need for stepping through the machine.
+	stepSize, err := p.computeStepSize(validatedHeights)
+	if err != nil {
+		return emptyCommit, err
+	}
+
+	// Compute how many machine hashes we need to collect.
+	numHashes, err := p.computeRequiredNumberOfHashes(validatedHeights, upToHeight)
+	if err != nil {
+		return emptyCommit, err
+	}
+
+	// Collect the machine hashes at the specified challenge level based on the values we computed.
+	hashes, err := p.machineHashCollector.CollectMachineMashes(
+		ctx,
+		&HashCollectorConfig{
+			WasmModuleRoot:    wasmModuleRoot,
+			MessageNumber:     Height(fromMessageNumber),
+			StepHeights:       validatedHeights[1:],
+			NumDesiredHashes:  numHashes,
+			MachineStartIndex: machineStartIndex,
+			StepSize:          stepSize,
+		},
+	)
+	if err != nil {
+		return emptyCommit, err
+	}
+	return commitments.New(hashes)
+}
+
+// Computes the required number of hashes for a history commitment
+// based on the requested heights and challenge level. The required number of hashes
+// for a leaf commitment at each challenge level is a constant, so we can determine
+// the desired challenge level from the input params and compute the total
+// from there.
+func (p *HistoryCommitmentProvider) computeRequiredNumberOfHashes(
+	startHeights validatedStartHeights,
+	upToHeight option.Option[Height],
+) (uint64, error) {
+	// Get the max number of hashes at the specified challenge level.
+	// from the protocol constants.
+	challengeLevel := len(startHeights) - 1
+	maxHeightForLevel := p.challengeLeafHeights[challengeLevel]
+
+	// Get the start height we want to use at the challenge level.
+	start := startHeights[challengeLevel]
+
+	var end Height
+	if upToHeight.IsNone() {
+		end = maxHeightForLevel
+	} else {
+		end = upToHeight.Unwrap()
+		// If the end height is more than the allowed max, we return an error.
+		// This scenario should not happen, and instead of silently truncating,
+		// surfacing an error is the safest way of warning the operator
+		// they are committing something invalid.
+		if end > maxHeightForLevel {
+			return 0, fmt.Errorf(
+				"end %d was greater than max height for level %d",
+				end,
+				maxHeightForLevel,
+			)
+		}
+	}
+	if end < start {
+		return 0, fmt.Errorf("invalid range: end %d was < start %d", end, start)
+	}
+	// The number of hashes is the difference between the start and end
+	// requested heights, plus 1.
+	return uint64(end-start) + 1, nil
```

Comment:
Given we dont hash for every opcode (height), don't we want to divide the step here?

Reply:
This is at a specific challenge level, so the number of hashes is just (end - start) + 1. For example, if I want a history commitment at the bigstep challenge level from height 0 to 1000, the number of hashes is just (1000 - 0) + 1 = 1001 hashes